les/txrelay.go

```diff
@@ -68,93 +55,92 @@ func (self *lesTxRelay) Stop() {
-			p, err := self.ps.getPeerWithEtherbase(tx.GatewayFeeRecipient())
-			// TODO(asa): When this happens, the nonce is still incremented, preventing future txs from being added.
-			// We rely on transactions to be rejected in light/txpool validateTx to prevent transactions
-			// with GatewayFeeRecipient != one of our peers from making it to the relayer.
-			if err != nil {
-				log.Error("Unable to find peer with matching etherbase", "err", err, "tx.hash", tx.Hash(), "tx.gatewayFeeRecipient", tx.GatewayFeeRecipient())
-				continue
-			}
-			sendTo[p] = append(sendTo[p], tx)
-			ltr := &ltrInfo{
-				tx:     tx,
-				sentTo: make(map[*peer]struct{}),
-			}
-			self.txSent[hash] = ltr
-			self.txPending[hash] = struct{}{}
+		if _, ok := self.txSent[hash]; ok {
+			continue
 		}
-	}
 
-	for p, list := range sendTo {
-		pp := p
-		ll := list
-		enc, _ := rlp.EncodeToBytes(ll)
+		self.txSent[hash] = tx
+		self.txPending[hash] = struct{}{}
 
+		// Send a single transaction per request to avoid failure coupling and
+		// because the expected base cost of a SendTxV2 request is 0, so it
+		// cost no extra to send multiple requests with one transaction each.
+		list := types.Transactions{tx}
+		enc, _ := rlp.EncodeToBytes(list)
+
+		// Assemble the request object with callbacks for the distributor.
 		reqID := genReqID()
 		rq := &distReq{
 			getCost: func(dp distPeer) uint64 {
-				peer := dp.(*peer)
-				return peer.GetTxRelayCost(len(ll), len(enc))
+				return dp.(*peer).GetTxRelayCost(len(list), len(enc))
 			},
 			canSend: func(dp distPeer) bool {
-				return !dp.(*peer).onlyAnnounce && dp.(*peer) == pp
+				return dp.(*peer).WillAcceptTransaction(tx)
 			},
 			request: func(dp distPeer) func() {
 				peer := dp.(*peer)
-				cost := peer.GetTxRelayCost(len(ll), len(enc))
+				cost := peer.GetTxRelayCost(len(list), len(enc))
 				peer.fcServer.QueuedRequest(reqID, cost)
 				return func() { peer.SendTxs(reqID, cost, enc) }
 			},
 		}
-		go self.retriever.retrieve(context.Background(), reqID, rq, func(p distPeer, msg *Msg) error { return nil }, self.stop)
+
+		// Check the response to see if the transaction was successfully added to the peer pool or mined.
+		// If an error is returned, the retreiver will retry with any remaining suitable peers.
+		checkTxStatus := func(p distPeer, msg *Msg) error {
+			if msg.MsgType != MsgTxStatus {
+				return errors.New("received unexpected messgae code")
+			}
+			statuses, ok := msg.Obj.([]light.TxStatus)
+			if !ok {
+				return errors.New("received invalid transaction status object")
+			}
+			if len(statuses) != 1 {
+				return errors.New("expected single transaction status response")
+			}
+			status := statuses[0]
+			if status.Error != "" {
+				return errors.New(status.Error)
+			}
+			if status.Status == core.TxStatusUnknown {
+				return errors.New("transaction status unknown")
+			}
+			return nil
+		}
+		go self.retriever.retrieve(context.Background(), reqID, rq, checkTxStatus, self.stop)
```

Comment:
`context.Background()` means we won't apply any kind of timeout logic to the request, are we sure about that?

Reply:
I've found this part of the code to be a bit of a rats nest as I have been diving into it. The original authors chose to pass through a context variable, but use a different mechanism for cancelations and timeouts. I looked at changing it with the PR, but decided it was too much to change for this piece of work, but I am still workin in this part of the code, so I can still change it in a following PR.